standard/expressions.md

```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
 
 A *primary_expression* that consists of a *literal* ([§7.4.5](lexical-structure.md#745-literals)) is classified as a value.
 
+### §interpolated-string-expressions Interpolated string expressions
+
+An *interpolated_string_expression* consists of a `$` character immediately followed by text within `"` characters. Within the quoted text there are zero or more **_interpolations_** delimited by `{` and `}` characters, each of which encloses an *expression* and optional formatting specifications.
+
+Interpolated string expressions have two forms; regular (*interpolated_regular_string_expression*)
+and verbatim (*interpolated_verbatim_string_expression*); which are lexically similar to, but differ semantically from, the two forms of string
+literals ([§7.4.5.6](lexical-structure.md#7456-string-literals)).
+
+```ANTLR
+interpolated_string_expression
+    : interpolated_regular_string_expression
+    | interpolated_verbatim_string_expression
+    ;
+
+// interpolated regular string expressions
+
+interpolated_regular_string_expression
+    : Interpolated_Regular_String_Start Interpolated_Regular_String_Mid?
+      ('{' regular_interpolation '}' Interpolated_Regular_String_Mid?)* Interpolated_Regular_String_End
+    ;
+
+regular_interpolation
+    : expression (',' interpolation_minimum_width)? Regular_Interpolation_Format?
+    ;
+
+interpolation_minimum_width
+    : constant_expression
+    ;
+
+Interpolated_Regular_String_Start
+    : '$"'
+    ;
+
+// the following three lexical rules are context sensitive, see details below
+Interpolated_Regular_String_Mid
+    : Interpolated_Regular_String_Element+
+    ;
+
+Regular_Interpolation_Format
+    : ':' Interpolated_Regular_String_Element+
+    ;
+
+Interpolated_Regular_String_End
+    : '"'
+    ;
+
+fragment Interpolated_Regular_String_Element
+    : Interpolated_Regular_String_Character
+    | Simple_Escape_Sequence
+    | Hexadecimal_Escape_Sequence
+    | Unicode_Escape_Sequence
+    | Open_Brace_Escape_Sequence
+    | Close_Brace_Escape_Sequence
+    ;
+
+fragment Interpolated_Regular_String_Character
+    // Any character except " (U+0022), \\ (U+005C), { (U+007B), } (U+007D), and New_Line_Character
+    : ~["\\{}\u000D\u000A\u0085\u2028\u2029]
+    ;
+
+// interpolated verbatim string expressions
+
+interpolated_verbatim_string_expression
+    : Interpolated_Verbatim_String_Start Interpolated_Verbatim_String_Mid?
+      ('{' verbatim_interpolation '}' Interpolated_Verbatim_String_Mid?)* Interpolated_Verbatim_String_End
+    ;
+
+verbatim_interpolation
+    : expression (',' interpolation_minimum_width)? Verbatim_Interpolation_Format?
+    ;
+
+Interpolated_Verbatim_String_Start
+    : '$@"'
+    ;
+
+// the following three lexical rules are context sensitive, see details below
+Interpolated_Verbatim_String_Mid
+    : Interpolated_Verbatim_String_Element+
+    ;
+
+Verbatim_Interpolation_Format
+    : ':' Interpolated_Verbatim_String_Element+
+    ;
+
+Interpolated_Verbatim_String_End
```

Comment:
Do we particularly need this to be separate from Interpolated_Regular_String_End? Could we just have Interpolated_String_End and use it in both rules? Or is that harder due to the context-sensitive nature?

Same applies for some other rules as well, I believe - it's just that this was the simplest one to validate.

Reply:
Correct, it is the context-sensitivity.

In the working ANTLR grammar there are two (in addition to the standard one) modes, one for each of regular & verbatim interpolated strings. In these two modes some rules are different while others are the same but with different names (as there is a single scope for names). This is a common situation when using modes, nothing special for the use case here.